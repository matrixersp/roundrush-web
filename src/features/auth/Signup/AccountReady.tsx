import { Stack, Typography, Link as MuiLink, styled } from '@mui/material';
import mail from 'icons/mail.svg';
import { Link } from 'react-router-dom';

export const AccountReady = () => {
  return (
    <ContainerStyled spacing={4.5}>
      <img src={mail} alt="mail icon" width="107" height="107" />
      <Typography>
        Please verify your email and continue to RoundRush
      </Typography>
      <MuiLink component={Link} to="/login">
        Go to login page
      </MuiLink>
    </ContainerStyled>
  );
};

const ContainerStyled = styled(Stack)(({ theme }) => ({
  alignItems: 'center',
  '& > .MuiTypography-body1': {
    lineHeight: 1.625,
    color: theme.palette.text.secondary,
    textAlign: 'center',
    maxWidth: 250,
    marginBottom: theme.spacing(3.75),
  },
}));
