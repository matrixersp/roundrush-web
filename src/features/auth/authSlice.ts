import { createAsyncThunk, createSlice } from '@reduxjs/toolkit';
import axios from 'axios';
import { BASE_API_URL } from 'utils/constants';

interface LoginData {
  email: string;
  password: string;
}

interface AuthError {
  error: string;
}
interface Auth {
  user: {
    _id: string;
    fullName: string;
    email: string;
  };
  token: string;
}

export const login = createAsyncThunk<Auth, LoginData>(
  'auth/login',
  async (data, { rejectWithValue }) => {
    try {
      const response = await axios.post(`${BASE_API_URL}/auth`, data);

      localStorage.setItem('TOKEN', response.data.token);
      return response.data as Auth;
    } catch (err) {
      if (axios.isAxiosError(err) && err.response) {
        return rejectWithValue(err.response.data.error as AuthError);
      }
      return rejectWithValue((err as Error).message);
    }
  }
);

export const verifyEmail = createAsyncThunk<Auth, string>(
  'auth/verifyEmail',
  async (email, { rejectWithValue }) => {
    try {
      const response = await axios.get(`${BASE_API_URL}/auth/verify-email`, {
        params: { email },
      });

      return response.data as Auth;
    } catch (err) {
      if (axios.isAxiosError(err) && err.response) {
        return rejectWithValue(err.response.data.error as AuthError);
      }
      return rejectWithValue((err as Error).message);
    }
  }
);

export const signup = createAsyncThunk<Auth, string>(
  'auth/signup',
  async (data, { rejectWithValue }) => {
    try {
      const response = await axios.post(`${BASE_API_URL}/users`, data);

      return response.data as Auth;
    } catch (err) {
      if (axios.isAxiosError(err) && err.response) {
        return rejectWithValue(err.response.data.error as AuthError);
      }
      return rejectWithValue((err as Error).message);
    }
  }
);

export const logout = createAsyncThunk('auth/logout', () => {
  localStorage.clear();
});

function isLoggedIn() {
  return localStorage.getItem('TOKEN') ? true : false;
}

type InitialState = {
  loading: boolean;
  loggedIn: boolean;
  error?: string;
};

const initialState: InitialState = {
  error: '',
  loading: false,
  loggedIn: isLoggedIn(),
};

export const authSlice = createSlice({
  name: 'auth',
  initialState,
  reducers: {},
  extraReducers: (builder) => {
    builder
      .addCase(login.fulfilled, (state, action) => {
        state.loading = false;
        state.loggedIn = true;
        state.error = '';
      })
      .addCase(login.pending, (state) => {
        state.loading = true;
      })
      .addCase(login.rejected, (state, action) => {
        state.loading = false;
        state.error = action.error.message as string;
      });

    builder
      .addCase(verifyEmail.fulfilled, (state, action) => {
        state.loading = false;
        state.error = '';
      })
      .addCase(verifyEmail.pending, (state) => {
        state.loading = true;
      })
      .addCase(verifyEmail.rejected, (state, action) => {
        state.loading = false;
        // state.error = action.error.message as string;
      });

    builder
      .addCase(signup.fulfilled, (state, action) => {
        state.loading = false;
        state.error = '';
      })
      .addCase(signup.pending, (state) => {
        state.loading = true;
      })
      .addCase(signup.rejected, (state, action) => {
        state.loading = false;
        state.error = action.error.message as string;
      });

    builder.addCase(logout.fulfilled, (state, action) => {
      state.loggedIn = false;
    });
  },
});

export default authSlice.reducer;
