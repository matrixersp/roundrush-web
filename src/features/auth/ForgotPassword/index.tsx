import { PublicLayout } from 'components/ui/PublicLayout';
import { ValidateEmail } from 'features/auth/ForgotPassword/ValidateEmail';
import { useLocation } from 'react-router-dom';
import { PasswordReset } from 'features/auth/ForgotPassword/PasswordReset';

export const ForgotPassword = () => {
  const location = useLocation();
  const token = new URLSearchParams(location.search).get('token');

  return (
    <PublicLayout
      description="Signing up a new account"
      title="Forgot your password"
    >
      {token ? <PasswordReset token={token} /> : <ValidateEmail />}
    </PublicLayout>
  );
};
