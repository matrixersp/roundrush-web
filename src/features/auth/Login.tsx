import { Stack, Link as MuiLink, styled, Button } from '@mui/material';
import { CustomTextField } from 'components/fields';
import { PublicLayout } from 'components/PublicLayout';
import { Formik, Form, Field } from 'formik';
import { Link, Navigate, useLocation, useNavigate } from 'react-router-dom';
import * as Yup from 'yup';
import { CustomAlert } from 'components/CustomAlert';
import { useState } from 'react';
import axios, { AxiosError } from 'axios';
import { useAppDispatch, useAppSelector } from 'app/hooks';
import { login } from 'features/auth/authSlice';
import { RootState } from 'app/store';

const loginSchema = Yup.object().shape({
  email: Yup.string().label('Email').required('${path} is required'),
  password: Yup.string().label('Password').required('${path} is required'),
});

export const Login = () => {
  const navigate = useNavigate();
  const location = useLocation();
  const dispatch = useAppDispatch();
  const isLoggedIn = useAppSelector(
    (state: RootState) => state.auth.isLoggedIn
  );
  const [error, setError] = useState('');

  const handleSubmit = async (values: any) => {
    dispatch(login(values))
      .unwrap()
      .then(() => {
        setError('');
        navigate(location.state?.from || '/dashboard');
      })
      .catch((err: AxiosError) => {
        if (axios.isAxiosError(err) && err.response) {
          setError(err.response.data.error || 'Something went wrong');
        } else {
          setError((err as AxiosError).message);
        }
      });
  };

  if (isLoggedIn) {
    return <Navigate to={location.state?.from || '/dashboard'} />;
  }

  return (
    <PublicLayout title="Login">
      <Formik
        validateOnBlur={false}
        initialValues={{
          email: '',
          password: '',
        }}
        validationSchema={loginSchema}
        onSubmit={async (values, actions) => {
          await handleSubmit(values);
          actions.setSubmitting(false);
          actions.resetForm();
        }}
      >
        {({ isSubmitting }) => (
          <Form>
            <Stack spacing={3} marginTop={6.375}>
              {error && (
                <CustomAlert severity="warning">
                  Please make sure you have the correct email and password
                </CustomAlert>
              )}
              {location.state?.reset && (
                <CustomAlert severity="success">
                  Well done. The password was reset successfully
                </CustomAlert>
              )}
              <Field
                autoFocus
                name="email"
                placeholder="Insert your email"
                type="text"
                component={CustomTextField}
                fullWidth
              />
              <Field
                name="password"
                placeholder="Password"
                type="password"
                component={CustomTextField}
                fullWidth
              />
              <ButtonStyled
                disabled={isSubmitting}
                type="submit"
                variant="contained"
                size="large"
              >
                Login
              </ButtonStyled>
              <LinksContainer gap={1}>
                <MuiLink component={Link} to="/forgot-password">
                  I forgot my password
                </MuiLink>
                <MuiLink component={Link} to="/signup">
                  I don't have an account
                </MuiLink>
              </LinksContainer>
            </Stack>
          </Form>
        )}
      </Formik>
    </PublicLayout>
  );
};

const ButtonStyled = styled(Button)(({ theme }) => ({
  paddingTop: theme.spacing(1.75),
  paddingBottom: theme.spacing(1.75),
}));

const LinksContainer = styled(Stack)(({ theme }) => ({
  justifyContent: 'space-between',
  flexDirection: 'row',
  [theme.breakpoints.down('sm')]: {
    flexDirection: 'column',
  },
}));
