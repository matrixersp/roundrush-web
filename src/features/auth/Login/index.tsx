import { Stack, Link as MuiLink, styled, Button } from '@mui/material';
import { CustomTextField } from 'components/fields';
import { PublicLayout } from 'components/ui/PublicLayout';
import { Formik, Form, Field, FormikHelpers } from 'formik';
import {
  Link,
  LinkProps,
  Navigate,
  useLocation,
  useNavigate,
} from 'react-router-dom';
import * as Yup from 'yup';
import { CustomAlert } from 'components/CustomAlert';
import { useAppDispatch, useAppSelector } from 'app/hooks';
import { login } from 'features/auth/authSlice';
import { RootState } from 'app/store';
import React from 'react';

const loginSchema = Yup.object().shape({
  email: Yup.string().email().label('Email').required('${path} is required'),
  password: Yup.string().label('Password').required('${path} is required'),
});

export const Login = () => {
  const navigate = useNavigate();
  const location = useLocation();
  const dispatch = useAppDispatch();
  const { loggedIn, error } = useAppSelector((state: RootState) => state.auth);

  const handleSubmit = async (values: any, actions: FormikHelpers<any>) => {
    dispatch(login(values))
      .then(() => {
        navigate(location.state?.from || '/dashboard');
      })
      .finally(() => {
        actions.setSubmitting(false);
        actions.resetForm();
      });
  };

  if (loggedIn) {
    return <Navigate to={location.state?.from || '/dashboard'} />;
  }

  return (
    <PublicLayout title="Login">
      <Formik
        initialValues={{
          email: '',
          password: '',
        }}
        validationSchema={loginSchema}
        onSubmit={handleSubmit}
      >
        {({ isSubmitting }) => (
          <Form aria-label="form">
            <Stack spacing={3} marginTop={6.375}>
              {error && (
                <CustomAlert severity="warning">
                  Please make sure you have the correct email and password
                </CustomAlert>
              )}
              {location.state?.reset && (
                <CustomAlert severity="success">
                  Well done. The password was reset successfully
                </CustomAlert>
              )}
              <Field
                autoFocus
                name="email"
                placeholder="Insert your email"
                type="text"
                component={CustomTextField}
                inputProps={{ 'data-testid': 'email' }}
                fullWidth
              />
              <Field
                name="password"
                placeholder="Password"
                type="password"
                component={CustomTextField}
                inputProps={{ 'data-testid': 'password' }}
                fullWidth
              />
              <ButtonStyled
                disabled={isSubmitting}
                type="submit"
                variant="contained"
                size="large"
              >
                Login
              </ButtonStyled>
              <LinksContainer gap={1}>
                <MuiLink
                  component={LinkBehavior}
                  to="/forgot-password"
                  data-testid="forgot-password-link"
                >
                  I forgot my password
                </MuiLink>
                <MuiLink
                  component={LinkBehavior}
                  to="/signup"
                  data-testid="register-link"
                >
                  I don't have an account
                </MuiLink>
              </LinksContainer>
            </Stack>
          </Form>
        )}
      </Formik>
    </PublicLayout>
  );
};

const LinkBehavior = React.forwardRef<any, LinkProps>(
  ({ to, ...props }, ref) => (
    <Link
      ref={ref}
      to={to}
      onMouseDown={(e) => e.preventDefault()}
      {...props}
    />
  )
);

const ButtonStyled = styled(Button)(({ theme }) => ({
  paddingTop: theme.spacing(1.75),
  paddingBottom: theme.spacing(1.75),
}));

const LinksContainer = styled(Stack)(({ theme }) => ({
  justifyContent: 'space-between',
  flexDirection: 'row',
  [theme.breakpoints.down('sm')]: {
    flexDirection: 'column',
  },
}));
