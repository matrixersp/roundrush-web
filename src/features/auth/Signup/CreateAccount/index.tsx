import { Stack, styled, Button, Typography } from '@mui/material';
import { Formik, Form, FormikHelpers, FormikProps } from 'formik';
import { Dispatch, SetStateAction, useEffect, useState } from 'react';
import { validationSchema } from 'features/auth/Signup/CreateAccount/validationSchema';
import { initialValues } from 'features/auth/Signup/CreateAccount/initialValues';
import { UserInformation } from 'features/auth/Signup/CreateAccount/UserInformation';
import { OrganizationInformation } from 'features/auth/Signup/CreateAccount/OrganizationInformation';
import { useAppDispatch } from 'app/hooks';
import { signup } from 'features/auth/authSlice';
import {
  getIndustries,
  getEmployeesSize,
} from 'features/organization/organizationSlice';

function renderStepContent(step: number, formikProps: FormikProps<any>) {
  switch (step) {
    case 0:
      return <UserInformation />;
    case 1:
      return <OrganizationInformation formikProps={formikProps} />;
    default:
      return null;
  }
}

type Props = {
  setAccountReady: Dispatch<SetStateAction<boolean>>;
  email: string;
};

export const CreateAccount = ({ setAccountReady, email }: Props) => {
  const [step, setStep] = useState(0);
  const formInitialValues = { ...initialValues, email };
  const currentValidationSchema = validationSchema[step];
  const isLastStep = step === 1;
  const dispatch = useAppDispatch();

  useEffect(() => {
    dispatch(getIndustries());
    dispatch(getEmployeesSize());
  }, []);

  const submitForm = async (values: any, actions: FormikHelpers<any>) => {
    dispatch(signup(values))
      .unwrap()
      .then(() => {
        setAccountReady(true);
      })
      .finally(() => {
        actions.setSubmitting(false);
      });
  };

  const handleSubmit = (values: any, actions: FormikHelpers<any>) => {
    if (isLastStep) {
      submitForm(values, actions);
    } else {
      setStep(step + 1);
      actions.setTouched({});
      actions.setSubmitting(false);
    }
  };

  return (
    <>
      <DescriptionStyled>Fill up your account information</DescriptionStyled>
      <Formik
        enableReinitialize={false}
        initialValues={formInitialValues}
        validationSchema={currentValidationSchema}
        onSubmit={handleSubmit}
      >
        {({ isSubmitting, ...props }) => (
          <Form>
            <Stack spacing={3}>
              {renderStepContent(step, { isSubmitting, ...props })}
              <ButtonStyled
                variant="contained"
                color={isLastStep ? 'secondary' : 'primary'}
                size="large"
                type="submit"
                disabled={isSubmitting}
              >
                {isLastStep ? 'Register' : 'Next'}
              </ButtonStyled>
            </Stack>
          </Form>
        )}
      </Formik>
    </>
  );
};

const DescriptionStyled = styled(Typography)(({ theme }) => ({
  color: theme.palette.text.secondary,
  marginBottom: theme.spacing(5),
  textAlign: 'center',
}));

const ButtonStyled = styled(Button)(({ theme }) => ({
  paddingTop: theme.spacing(1.75),
  paddingBottom: theme.spacing(1.75),
}));
