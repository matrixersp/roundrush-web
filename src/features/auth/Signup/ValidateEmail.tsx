import { Stack, styled, Button, Typography } from '@mui/material';
import { CustomTextField } from 'components/fields';
import { Formik, Form, Field, FormikHelpers } from 'formik';
import * as Yup from 'yup';
import { useState } from 'react';
import { Link } from 'react-router-dom';
import { verifyEmail } from 'features/auth/authSlice';
import { useAppDispatch } from 'app/hooks';

const signupSchema = Yup.object().shape({
  email: Yup.string().email().label('Email').required('${path} is required'),
});

type Props = {
  onContinue: (email: string) => void;
};

export const ValidateEmail = ({ onContinue }: Props) => {
  const [step, setStep] = useState(0);
  const dispatch = useAppDispatch();

  const handleSubmit = async ({ email }: any, actions: FormikHelpers<any>) => {
    dispatch(verifyEmail(email))
      .unwrap()
      .then(() => {
        onContinue(email);
      })
      .catch(() => {
        setStep(1);
      })
      .finally(() => {
        actions.setSubmitting(false);
        actions.resetForm();
      });
  };

  return step === 0 ? (
    <>
      <DescriptionStyled>Letâ€™s validate your email first</DescriptionStyled>
      <Formik
        initialValues={{ email: '' }}
        validationSchema={signupSchema}
        onSubmit={handleSubmit}
      >
        {({ isSubmitting }) => (
          <Form>
            <Stack spacing={3}>
              <Field
                autoFocus
                name="email"
                placeholder="Insert your email"
                type="text"
                component={CustomTextField}
                fullWidth
              />
              <ButtonStyled
                disabled={isSubmitting}
                type="submit"
                variant="contained"
                size="large"
              >
                Next
              </ButtonStyled>
            </Stack>
          </Form>
        )}
      </Formik>
    </>
  ) : (
    <>
      <DescriptionStyled>
        Your email already exists in RoundRush
      </DescriptionStyled>
      <ButtonStyled
        // @ts-expect-error
        component={Link}
        to="/login"
        variant="contained"
        size="large"
        fullWidth
      >
        Click here to authenticate
      </ButtonStyled>
    </>
  );
};

const DescriptionStyled = styled(Typography)(({ theme }) => ({
  color: theme.palette.text.secondary,
  marginBottom: theme.spacing(5),
  textAlign: 'center',
}));

const ButtonStyled = styled(Button)(({ theme }) => ({
  paddingTop: theme.spacing(1.75),
  paddingBottom: theme.spacing(1.75),
}));
