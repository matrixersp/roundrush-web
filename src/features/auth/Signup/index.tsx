import { PublicLayout } from 'components/ui/PublicLayout';
import { CreateAccount } from 'features/auth/Signup/CreateAccount';
import { ValidateEmail } from 'features/auth/Signup/ValidateEmail';
import { AccountReady } from 'features/auth/Signup/AccountReady';
import { useState } from 'react';

export const Signup = () => {
  const [email, setEmail] = useState('');
  const [accountReady, setAccountReady] = useState(false);

  return (
    <PublicLayout
      description="Signing up a new account"
      title={accountReady ? 'Your account is ready!' : 'Sign up'}
    >
      {accountReady ? (
        <AccountReady />
      ) : email ? (
        <CreateAccount email={email} setAccountReady={setAccountReady} />
      ) : (
        <ValidateEmail onContinue={(email) => setEmail(email)} />
      )}
    </PublicLayout>
  );
};
