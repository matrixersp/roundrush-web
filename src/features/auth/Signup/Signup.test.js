import { screen, within } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { rest } from 'msw';
import { setupServer } from 'msw/lib/node';
import { renderWithProviders } from 'utils/test-utils';
import { Signup } from 'features/auth/Signup';
import { handlers } from 'utils/test-handlers';
import { BASE_API_URL } from 'utils/constants';

const server = setupServer(...handlers);

beforeAll(() => server.listen());
afterEach(() => server.resetHandlers());
afterAll(() => server.close());

describe('Signup', () => {
  it('should render correctly', () => {
    const { container } = renderWithProviders(<Signup />);
    expect(container).toMatchSnapshot();
  });
});

describe('ValidateEmail', () => {
  it('should render a form with 1 input and a button', () => {
    renderWithProviders(<Signup />);

    expect(
      screen.getByPlaceholderText('Insert your email')
    ).toBeInTheDocument();
    expect(screen.getByText('Next')).toBeInTheDocument();
  });

  it('should render an error message when the email is invalid', async () => {
    renderWithProviders(<Signup />);

    const input = screen.getByPlaceholderText('Insert your email');
    const button = screen.getByText('Next');

    await userEvent.click(button);
    expect(await screen.findByText('Email is required')).toBeInTheDocument();

    await userEvent.type(input, 'invalid-email');
    await userEvent.click(button);

    expect(
      await screen.findByText('Email must be a valid email')
    ).toBeInTheDocument();
  });

  it('should display a button to authenticate when email already exists', async () => {
    server.use(
      rest.get(`${BASE_API_URL}/auth/verify-email`, async (req, res, ctx) => {
        return res(
          ctx.status(400),
          ctx.json({
            message: 'Email already exists',
          })
        );
      })
    );

    renderWithProviders(<Signup />);

    const input = screen.getByPlaceholderText('Insert your email');
    const button = screen.getByText('Next');

    await userEvent.type(input, 'joe@example.com');
    await userEvent.click(button);

    expect(
      await screen.findByText('Click here to authenticate')
    ).toBeInTheDocument();
    expect(input).not.toBeInTheDocument();
    expect(button).not.toBeInTheDocument();
  });

  it('should display a user information form when email does not exist', async () => {
    renderWithProviders(<Signup />);

    const input = screen.getByPlaceholderText('Insert your email');
    const button = screen.getByText('Next');

    await userEvent.type(input, 'joe2@example.com');
    await userEvent.click(button);

    expect(await screen.findByPlaceholderText('Full name')).toBeInTheDocument();
    expect(await screen.findByPlaceholderText('Password')).toBeInTheDocument();
    expect(await screen.findByText('Next')).toBeInTheDocument();

    expect(input).not.toBeInTheDocument();
    expect(button).not.toBeInTheDocument();
  });
});

describe('UserInformation', () => {
  const setup = async () => {
    renderWithProviders(<Signup />);
    const input = screen.getByPlaceholderText('Insert your email');
    const button = screen.getByText('Next');

    await userEvent.type(input, 'joe2@example.com');
    await userEvent.click(button);

    expect(await screen.findByPlaceholderText('Full name')).toBeInTheDocument();
    expect(await screen.findByPlaceholderText('Password')).toBeInTheDocument();
    expect(await screen.findByText('Next')).toBeInTheDocument();

    expect(input).not.toBeInTheDocument();
    expect(button).not.toBeInTheDocument();
  };

  it('should render a form with 2 inputs and a button', async () => {
    await setup();
    const nameInput = screen.getByPlaceholderText('Full name');
    const passwordInput = screen.getByPlaceholderText('Password');
    const button = screen.getByText('Next');

    expect(nameInput).toBeInTheDocument();
    expect(passwordInput).toBeInTheDocument();
    expect(button).toBeInTheDocument();
  });

  it('should display an error message when the name or password are invalid', async () => {
    await setup();
    const nameInput = await screen.findByPlaceholderText('Full name');
    const passwordInput = await screen.findByPlaceholderText('Password');
    const nextButton = await screen.findByText('Next');

    await userEvent.click(nextButton);

    expect(
      await screen.findByText('Full name is required')
    ).toBeInTheDocument();
    expect(await screen.findByText('Password is required')).toBeInTheDocument();

    await userEvent.type(nameInput, 'J');
    await userEvent.type(passwordInput, '123');
    await userEvent.click(nextButton);

    expect(
      await screen.findByText('Full name must be at least 2 characters')
    ).toBeInTheDocument();
    expect(
      await screen.findByText('Password must be at least 8 characters')
    ).toBeInTheDocument();
  });

  it('should display organization form when name and password are valid', async () => {
    await setup();
    const nameInput = await screen.findByPlaceholderText('Full name');
    const passwordInput = await screen.findByPlaceholderText('Password');
    const nextButton = await screen.findByText('Next');

    await userEvent.type(nameInput, 'Joe Doe');
    await userEvent.type(passwordInput, '12345678');
    await userEvent.click(nextButton);

    expect(
      await screen.findByPlaceholderText('Company name')
    ).toBeInTheDocument();
    expect(await screen.findByPlaceholderText('Space')).toBeInTheDocument();
    expect(await screen.findByText('Industry type')).toBeInTheDocument();
    expect(await screen.findByText('Select a size..')).toBeInTheDocument();
    expect(await screen.findByText('Register')).toBeInTheDocument();

    expect(nameInput).not.toBeInTheDocument();
    expect(passwordInput).not.toBeInTheDocument();
    expect(screen.queryByText('Next')).not.toBeInTheDocument();
  });
});

describe('OrganizationInformation', () => {
  const setup = async () => {
    renderWithProviders(<Signup />);
    const input = screen.getByPlaceholderText('Insert your email');
    const button = screen.getByText('Next');

    await userEvent.type(input, 'joe@example.com');
    await userEvent.click(button);

    expect(await screen.findByPlaceholderText('Full name')).toBeInTheDocument();
    expect(await screen.findByPlaceholderText('Password')).toBeInTheDocument();
    expect(await screen.findByText('Next')).toBeInTheDocument();

    expect(input).not.toBeInTheDocument();
    expect(button).not.toBeInTheDocument();

    const nameInput = await screen.findByPlaceholderText('Full name');
    const passwordInput = await screen.findByPlaceholderText('Password');
    const nextButton = await screen.findByText('Next');

    await userEvent.type(nameInput, 'Joe Doe');
    await userEvent.type(passwordInput, '12345678');
    await userEvent.click(nextButton);

    expect(
      await screen.findByPlaceholderText('Company name')
    ).toBeInTheDocument();
    expect(await screen.findByPlaceholderText('Space')).toBeInTheDocument();
    expect(await screen.findByText('Industry type')).toBeInTheDocument();
    expect(await screen.findByText('Select a size..')).toBeInTheDocument();
    expect(await screen.findByText('Register')).toBeInTheDocument();

    expect(nameInput).not.toBeInTheDocument();
    expect(passwordInput).not.toBeInTheDocument();
    expect(screen.queryByText('Next')).not.toBeInTheDocument();
  };

  it('should render a form with 4 inputs and a button', async () => {
    await setup();
    const companyNameInput = screen.getByPlaceholderText('Company name');
    const spaceInput = screen.getByPlaceholderText('Space');
    const industryTypeInput = screen.getByText('Industry type');
    const sizeInput = screen.getByText('Select a size..');
    const registerButton = screen.getByText('Register');

    expect(companyNameInput).toBeInTheDocument();
    expect(spaceInput).toBeInTheDocument();
    expect(industryTypeInput).toBeInTheDocument();
    expect(sizeInput).toBeInTheDocument();
    expect(registerButton).toBeInTheDocument();
  });

  it('should display an error message when the company name, space and terms are invalid', async () => {
    await setup();
    const companyNameInput = await screen.findByPlaceholderText('Company name');
    const spaceInput = await screen.findByPlaceholderText('Space');
    const industryTypeInput = await screen.findByTestId('industry-type');
    const registerButton = await screen.findByText('Register');

    await userEvent.click(registerButton);

    expect(
      await screen.findByText('Company name is required')
    ).toBeInTheDocument();
    expect(await screen.findByText('Space is required')).toBeInTheDocument();
    expect(
      await screen.findByText(
        'You must accept the Terms of Use and Privacy Policy'
      )
    ).toBeInTheDocument();

    await userEvent.type(companyNameInput, 'J');
    await userEvent.type(spaceInput, 'S');
    await userEvent.click(industryTypeInput);
    userEvent.click(
      await within(industryTypeInput).findByText('Industry type')
    );
    // userEvent.click(
    //   await screen.findByRole('option', { name: 'Industry type' })
    // );
    // await userEvent.selectOptions(
    //   industryTypeInput,
    //   screen.getByRole('option', { name: 'Agriculture' })
    // );
    await userEvent.click(registerButton);

    expect(
      await screen.findByText('Company name must be at least 2 characters')
    ).toBeInTheDocument();
    expect(
      await screen.findByText('Space must be at least 2 characters')
    ).toBeInTheDocument();
    expect(
      await screen.findByText(
        'You must accept the Terms of Use and Privacy Policy'
      )
    ).toBeInTheDocument();
  });

  it('should register the user when the form is valid', async () => {
    await setup();
    const companyNameInput = await screen.findByPlaceholderText('Company name');
    const spaceInput = await screen.findByPlaceholderText('Space');
    const industryTypeInput = await screen.findByTestId('industry-type');
    const sizeInput = await screen.findByTestId('employees-size');
    const registerButton = await screen.findByText('Register');
    const termsCheckbox = await screen.findByTestId('accept-terms');

    await userEvent.type(companyNameInput, 'Company');
    await userEvent.type(spaceInput, 'Space');
    // await userEvent.click(industryTypeInput);
    // userEvent.click(await within(industryTypeInput).findByText('Agriculture'));
    // await userEvent.click(sizeInput);
    // userEvent.click(await within(sizeInput).findByText('1-10'));
    await userEvent.click(termsCheckbox);
    await userEvent.click(registerButton);

    expect(
      await screen.findByText('Your account is ready!')
    ).toBeInTheDocument();
    expect(
      await screen.findByRole('link', { name: 'Go to login page' })
    ).toBeInTheDocument();
  });
});
