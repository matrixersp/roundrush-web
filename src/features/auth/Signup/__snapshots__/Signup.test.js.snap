// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Signup should render correctly 1`] = `
<div>
  <div
    class="css-11iocy1"
  >
    <div
      class="css-uhvysr"
    />
    <img
      alt="Brand"
      src="brand.svg"
    />
    <p
      class="MuiTypography-root MuiTypography-body1 css-r0herm-MuiTypography-root"
    >
      Signing up a new account
    </p>
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation0 css-ee20y-MuiPaper-root"
    >
      <h1
        class="MuiTypography-root MuiTypography-h1 css-10yptsx-MuiTypography-root"
        data-testid="Sign up"
      >
        Sign up
      </h1>
      <p
        class="MuiTypography-root MuiTypography-body1 css-1najbws-MuiTypography-root"
      >
        Let’s validate your email first
      </p>
      <form
        action="#"
      >
        <div
          class="css-1lwbda4-MuiStack-root"
        >
          <div
            class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
          >
            <div
              class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth Mui-focused MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
            >
              <input
                aria-invalid="false"
                class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                id=":r0:"
                name="email"
                placeholder="Insert your email"
                type="text"
                value=""
              />
              <fieldset
                aria-hidden="true"
                class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
              >
                <legend
                  class="css-ihdtdm"
                >
                  <span
                    class="notranslate"
                  >
                    ​
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeLarge MuiButton-containedSizeLarge MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeLarge MuiButton-containedSizeLarge css-1e5b16l-MuiButtonBase-root-MuiButton-root"
            tabindex="0"
            type="submit"
          >
            Next
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
      </form>
    </div>
  </div>
</div>
`;
